### ISSUE TITLE
Add support for kimi agent

### ISSUE THREAD

**Issue by kamilio** [poe-platform/poe-code#23]
Write a spec doc how this could be achieved

# Kimi CLI

[![Commit Activity](https://img.shields.io/github/commit-activity/w/MoonshotAI/kimi-cli)](https://github.com/MoonshotAI/kimi-cli/graphs/commit-activity)
[![Checks](https://img.shields.io/github/check-runs/MoonshotAI/kimi-cli/main)](https://github.com/MoonshotAI/kimi-cli/actions)
[![Version](https://img.shields.io/pypi/v/kimi-cli)](https://pypi.org/project/kimi-cli/)
[![Downloads](https://img.shields.io/pypi/dw/kimi-cli)](https://pypistats.org/packages/kimi-cli)
[![Ask DeepWiki](https://deepwiki.com/badge.svg)](https://deepwiki.com/MoonshotAI/kimi-cli)

[中文](https://www.kimi.com/coding/docs/kimi-cli.html)

Kimi CLI is a new CLI agent that can help you with your software development tasks and terminal operations.

> [!IMPORTANT]
> Kimi CLI is currently in technical preview.

## Key features

- Shell-like UI and shell command execution
- Zsh integration
- [Agent Client Protocol] support
- MCP support
- And more to come...

[Agent Client Protocol]: https://github.com/agentclientprotocol/agent-client-protocol

## Installation

Kimi CLI is published as a Python package on PyPI. We highly recommend installing it with [uv](https://docs.astral.sh/uv/). If you have not installed uv yet, please follow the instructions [here](https://docs.astral.sh/uv/getting-started/installation/) to install it first.

Once uv is installed, you can install Kimi CLI with:

```sh
uv tool install --python 3.13 kimi-cli
```

Run `kimi --help` to check if Kimi CLI is installed successfully.

> [!IMPORTANT]
> Due to the security checks on macOS, the first time you run `kimi` command may take 10 seconds or more depending on your system environment.

## Upgrading

Upgrade Kimi CLI to the latest version with:

```sh
uv tool upgrade kimi-cli --no-cache
```

## Usage

Run `kimi` command in the directory you want to work on, then send `/setup` to setup Kimi CLI:

![](./docs/images/setup.png)

After setup, Kimi CLI will be ready to use. You can send `/help` to get more information.

### Shell mode

Kimi CLI is not only a coding agent, but also a shell. You can switch the mode by pressing `Ctrl-X`. In shell mode, you can directly run shell commands without leaving Kimi CLI.

> [!NOTE]
> Built-in shell commands like `cd` are not supported yet.

### Zsh integration

You can use Kimi CLI together with Zsh, to empower your shell experience with AI agent capabilities.

Install the [zsh-kimi-cli](https://github.com/MoonshotAI/zsh-kimi-cli) plugin via:

```sh
git clone https://github.com/MoonshotAI/zsh-kimi-cli.git \
  ${ZSH_CUSTOM:-~/.oh-my-zsh/custom}/plugins/kimi-cli
```

> [!NOTE]
> If you are using a plugin manager other than Oh My Zsh, you may need to refer to the plugin's README for installation instructions.

Then add `kimi-cli` to your Zsh plugin list in `~/.zshrc`:

```sh
plugins=(... kimi-cli)
```

After restarting Zsh, you can switch to agent mode by pressing `Ctrl-X`.

### ACP support

Kimi CLI supports [Agent Client Protocol] out of the box. You can use it together with any ACP-compatible editor or IDE.

For example, to use Kimi CLI with [Zed](https://zed.dev/), add the following configuration to your `~/.config/zed/settings.json`:

```json
{
  "agent_servers": {
    "Kimi CLI": {
      "command": "kimi",
      "args": ["--acp"],
      "env": {}
    }
  }
}
```

Then you can create Kimi CLI threads in Zed's agent panel.

### Using MCP tools

Kimi CLI supports the well-established MCP config convention. For example:

```json
{
  "mcpServers": {
    "context7": {
      "url": "https://mcp.context7.com/mcp",
      "headers": {
        "CONTEXT7_API_KEY": "YOUR_API_KEY"
      }
    },
    "chrome-devtools": {
      "command": "npx",
      "args": ["-y", "chrome-devtools-mcp@latest"]
    }
  }
}
```

Run `kimi` with `--mcp-config-file` option to connect to the specified MCP servers:

```sh
kimi --mcp-config-file /path/to/mcp.json
```

## Development

To develop Kimi CLI, run:

```sh
git clone https://github.com/MoonshotAI/kimi-cli.git
cd kimi-cli

make prepare  # prepare the development environment
```

Then you can start working on Kimi CLI.

Refer to the following commands after you make changes:

```sh
uv run kimi  # run Kimi CLI

make format  # format code
make check  # run linting and type checking
make test  # run tests
make help  # show all make targets
```

## Contributing

We welcome contributions to Kimi CLI! Please refer to [CONTRIBUTING.md](./CONTRIBUTING.md) for more information.

---

Kimi models on Poe

Kimi-K2
Kimi K2-Instruct-0905 is a state-of-the-art mixture-of-experts (MoE) language model with 32 billion activated parameters and 1 trillion total parameters. Trained with the Muon optimizer, Kimi K2 achieves exceptional performance across frontier knowledge, reasoning, and coding tasks while being meticulously optimized for agentic capabilities. Key Features: - Large-Scale Training: Pre-trained a 1T parameter MoE model on 15.5T tokens with zero training instability. - MuonClip Optimizer: We apply the Muon optimizer to an unprecedented scale, and develop novel optimization techniques to resolve instabilities while scaling up. - Agentic Intelligence: Specifically designed for tool use, reasoning, and autonomous problem-solving. Technical Specifications File Support: Attachments not supported Context window: 256k tokens
OFFICIAL

Kimi-K2-Thinking
Built as a thinking agent, it performs step-by-step reasoning while utilizing tools, achieving state-of-the-art performance on benchmarks such as Humanity's Last Exam (HLE), BrowseComp, and others. The model demonstrates substantial advancements in reasoning, agentic search, coding, writing, and general problem-solving capabilities. Kimi K2 Thinking is capable of executing 200–300 sequential tool calls autonomously, maintaining coherent reasoning across hundreds of steps to solve complex tasks. File Support: Text, Markdown and PDF files Context window: 256k tokens
OFFICIAL

Kimi-K2-T
Kimi K2 is a state-of-the-art mixture-of-experts (MoE) language model with 32 billion activated parameters and 1 trillion total parameters. Trained with the Muon optimizer, Kimi K2 achieves exceptional performance across frontier knowledge, reasoning, and coding tasks while being meticulously optimized for agentic capabilities.
OFFICIAL

Kimi-K2-Instruct
Kimi K2 is a state-of-the-art mixture-of-experts (MoE) language model with 32 billion activated parameters and 1 trillion total parameters. Trained with the Muon optimizer, Kimi K2 achieves exceptional performance across frontier knowledge, reasoning, and coding tasks while being meticulously optimized for agentic capabilities. Uses the latest September 5th, 2025 snapshot. The updated version has improved coding abilities, agentic tool use, and a longer (256K) context window.
OFFICIAL

Kimi-K2-0905-T
The new Kimi K2-0905 model from Moonshot AI features a massive 256,000-token context window, double the length of its predecessor (Kimi K2), along with greatly improved coding abilities and front-end generation accuracy. It boasts 1 trillion total parameters (with 32 billion activated at a time) and claims 100% tool-call success in real-world tests, setting a new bar for open-source AI performance in complex, multi-step tasks
OFFICIAL

Kimi-K2-Think-T
Kimi K2 Thinking is Moonshot AI's most capable open-source thinking model, built as a thinking agent that reasons step-by-step while dynamically invoking tools. Setting new state-of-the-art records on Humanity's Last Exam (HLE), BrowseComp, and other benchmarks, K2 Thinking dramatically scales multi-step reasoning depth while maintaining stable tool-use across 200–300 sequential calls — a breakthrough in long-horizon agency with native INT4 quantization for 2x inference speed. Supported File Types: JPEG, PNG, PDF
OFFICIAL

---

I'll look through the repository to find where the CLI reads configuration for base URL and model selection (config files, environment variables, or CLI flags), then I'll explain how to set a custom base URL and choose different models. I'll search the repo for relevant files and code paths. Now I'll query the repository.You can configure custom base URLs and multiple/different models in two ways: edit the CLI config file (~/.kimi/config.json) or set up programmatically (or via the interactive setup). The repo stores provider and model definitions in the top-level Config object:

- providers: mapping of provider-name -> LLMProvider (type, base_url, api_key, custom_headers)
- models: mapping of model-id -> LLMModel (provider, model, max_context_size, capabilities)
- default_model: model-id string (must exist in models)

Important constraints
- default_model (if set) must be a key in models.
- each LLMModel.provider must be a key present in providers.

1) Edit the config file (~/.kimi/config.json)
The config file path used by the CLI is ~/.kimi/config.json (see README / AGENTS.md). Example minimal config JSON that adds a custom provider with a custom base URL and two models (one set as default):

```json name=.kimi/config.json
{
  "default_model": "kimi-k2-turbo-preview",
  "models": {
    "kimi-k2-turbo-preview": {
      "provider": "kimi",
      "model": "kimi-k2-turbo-preview",
      "max_context_size": 250000
    },
    "kimi-k2-small": {
      "provider": "kimi",
      "model": "kimi-k2-small",
      "max_context_size": 100000
    }
  },
  "providers": {
    "kimi": {
      "type": "kimi",
      "base_url": "https://api.moonshot.ai/v1",
      "api_key": "YOUR_API_KEY_HERE",
      "custom_headers": {
        "X-Custom": "value"
      }
    },
    "my-other-provider": {
      "type": "kimi",
      "base_url": "https://my.custom-host.example/api/v1",
      "api_key": "ANOTHER_API_KEY"
    }
  },
  "loop_control": {
    "max_steps_per_run": 100,
    "max_retries_per_step": 3
  },
  "services": {
    "moonshot_search": {
      "base_url": "https://api.moonshot.ai/v1/search",
      "api_key": "YOUR_API_KEY"
    },
    "moonshot_fetch": {
      "base_url": "https://api.moonshot.ai/v1/fetch",
      "api_key": "YOUR_API_KEY"
    }
  }
}
```

Notes:
- Replace the api_key values with actual secrets (the CLI stores them as strings in JSON; code uses SecretStr).
- custom_headers is optional.
- You can add multiple provider entries if you want more than one backend (e.g., different environments, testing vs prod).
- max_context_size should reflect what the model/provider supports.

2) Use the interactive setup (CLI) to create a provider + model
The interactive "setup" command in the shell will create a provider and model entries for you (it writes the provider and model and sets default_model). Use the project's shell/setup flow (the repo includes a meta command that registers platform id, base_url, and model). That is handy if you prefer not to edit JSON manually.

3) Programmatic creation / custom usage (example)
If you create an LLM instance in code you can pass a provider with any base_url and a model definition directly (the repo's examples show this pattern). Example:

```python name=examples/custom-llm-usage.py
from pydantic import SecretStr
from kimi_cli.config import LLMProvider, LLMModel
from kimi_cli.llm import create_llm

# Create provider and model programmatically
provider = LLMProvider(
    type="kimi",
    base_url="https://api.moonshot.ai/v1",     # <-- custom base url
    api_key=SecretStr("YOUR_API_KEY")
)

model = LLMModel(
    provider="kimi",
    model="kimi-k2-turbo-preview",             # <-- which model name to use at provider
    max_context_size=250_000
)

llm = create_llm(provider, model)
# pass llm into your soul/agent/session as shown in examples/
```

You can also set base_url and api_key via environment variables (see the examples in the repo) and use that when creating providers programmatically.

4) Switching models at runtime
- To change the default used by the CLI, update "default_model" in the config to a different model key.
- Programmatically, when instantiating a soul/LLM you can pass a different LLMModel.
- If you need multiple models available simultaneously, add them all into the "models" mapping and reference the desired key.

5) Services (search/fetch) custom URLs
If you rely on repository services (moonshot_search / moonshot_fetch), add their base_url and api_key in the services section of the config (see example above). The CLI will use these service endpoints where applicable.

Quick checklist
- Add provider with base_url and api_key under providers
- Add model(s) under models and point their provider field to a provider key
- Set default_model to one of the model keys (optional)
- Restart the CLI / re-load config if necessary

If you want, tell me whether you prefer to edit ~/.kimi/config.json manually or use the interactive setup and I can give a ready-to-copy config snippet tuned to your provider URLs and models.
